# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
#regular spegular function

def addTwo(num):
    return num + 2

print(addTwo(4))

#Lambda Syntax for defining a function without saving it to a variable
print((lambda num: num + 2)(4))

6
6


#### Saving to a Variable

In [3]:
#we can save lambdas as variables
#its almost like saving a function
#like list comprehension for function

f_test = lambda num: num + 2

f_test(4)

#if you were to wrap f_test(4) in a print(f_test(4)), you wouldn't see Out[3]: 6 in your results

6

#### Multiple Inputs

In [5]:
#3 ways to do this:
#1: lambda function multiple inputs with no variable 
print((lambda x,y,z: x * y * z)(3, 5, 8))

#2: longway of writing this function

def multiply_nums(x,y,z):
    return x * y * z
print(multiply_nums(3,5,8))

#3: Lambda function with multiple inputs saved to a variable

num_test = lambda x,y,z: x * y * z
print(num_test(3,5,8))

120
120
120


#### Passing a Lambda into a Function

In [9]:
def multiply(func, num):
    """
    func expects lambda function
    and num expects and integer
    """
    return func(num)
print(multiply(lambda x: x * x, 4))

16


#### Returning a Lambda from a Function

In [18]:
#regular way of defining function

def multiply_test(num):
    return num * 4 
print(multiply_test(2))

#syntax for writing function inside of funtions

def return_func():
    test = 4
    def mulitply(num):
        return num * 4
    return multiply

f_return = return_func()
print(f_return(2))

#Like to use lambdas instead inside functions
def return_lam(a, b):
    
    return lambda c, d: c + d + a + b
r_return = return_lamb(3,4)
print(r_return)
print(r_return(5,6))

8


TypeError: multiply() missing 1 required positional argument: 'num'

In [22]:
#will find lambdas inside of functions
def my_func(num1, num2):
    print(num1, num2)
    
    return lambda num3, num4: (num3 * num4) + (num1 * num2)

print("with saving to variable")
aften_var = my_func(1,2)
print(aften_var)
print(aften_var(3,4))

print("\nWithout saving to variable")
print(my_func(1,2)(3,4))

with saving to variable
1 2
<function my_func.<locals>.<lambda> at 0x7f7f482095e0>
14

Without saving to variable
1 2
14


#### If Statements within Lambdas

In [24]:
#syntax for "if statements" in lambdas
#lambda x: True if (condition) else False

f_conditional = lambda num: num * 2 if num > 10 else num + 2

print(f_conditional(8))
print(f_conditional(10))
print(f_conditional(12))


10
12
24


In [32]:
#example of an if, elif, else using lambdas

alex_examp = lambda num: f"{num} is less than 5" if num < 5 else\
(f"{num} is greater than 5 and less than 10" if num < 10 else f"{num} is greater than 10")

print(alex_examp(3))
print(alex_examp(8))
print(alex_examp(12))

3 is less than 5
8 is greater than 5 and less than 10
12 is greater than 10


In [34]:
alex_examp = lambda num: f"{num} is less than 5" if num < 5 else\
(f"{num} is greater than 5 and less than 10" if num < 10 else\
 (f"{num} is greater than 10 and less than 20" if num < 20 else f"{num} is greater than 20"))

print(alex_examp(3))
print(alex_examp(8))
print(alex_examp(12))
print(alex_examp(22))

3 is less than 5
8 is greater than 5 and less than 10
12 is greater than 10 and less than 20
22 is greater than 20


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [48]:
print((lambda x, y, z: x * x * x)(5,5,5))
print((lambda x, y, z: y * y * y)(6,6,6))
print((lambda x, y, z: z * z * z)(7,7,7))

125
216
343


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [53]:
#syntax for utilizing map function
#map(functions, iterable(list, dictionaries, tuples, sets))
#normally the usage of map happens with a predefined function -- but we can use lambdas

#utilizing the map function with a regular pre-defined user function

def squared(num1, num2):
    
    if num1 < 10 and num2 < 10: 
        return num1 ** 2, num2 ** 2
    else:
        return num1, num2
    
numbers = [4, 11, 20, 3, 15, 20]
more_nums = [4, 10, 3, 2, 6]

squared_nums = list(map(squared, numbers, more_nums))
print(squared_nums)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [57]:
#syntax for map and lambdas
#list(map(lambda x: (transformation on x), list (or other iterable)))
#usually happens in one link so it's cleaner code

squared_nums_again = list(map(lambda num1, num2: (num1 ** 2, num2 ** 2) if num1 < 10 and num2 < 10\
                              else (num1, num2), numbers, more_nums))
print(squared_nums_again)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [64]:
my_list = [2, 4, 6, 8, 10, 12, 14, 16]

new_list = list(map(lambda x: x * 2 - 1 , my_list))

print(new_list)

[3, 7, 11, 15, 19, 23, 27, 31]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [66]:
#syntax for filter
#list(filter(function, iterable(list, dict tuples, set, etc)))

names = ["Chris", "Steve", "Perry", "Kelsey"]

def y_names(names):
    
    if names[-1].lower() == "y":
        return True
    else:
        return False
new_names_list = list(filter(y_names, names))

#for elephant in names:
#if y_names(elephant) == True:

print(new_names_list)

['Perry', 'Kelsey']


#### Using Lambda's with Filter()

In [67]:
#syntax for filter
#list(filter(function, iterable(list, dict, tip;es, set etc)))

new_names_lamb = list(filter(lambda name: True if name[-1].lower() == "y" else False, names))
print(new_names_lamb)

['Perry', 'Kelsey']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [78]:
from statistics import mean
  
def Average(lst):
    return mean(lst)
  
lst = [15, 9, 55, 41, 35, 20, 62, 49]
average = Average(lst)
  
print("Average of the list =", round(average, 2))

Average of the list = 35.75


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [81]:
from functools import reduce

#syntax for reduce
#reduce(function, iterable)

list_1 = [2, 4, 6, 8, 10]

#using a user-defined function with reduce

def add_nums(num1, num2):
    return num1 + num2 
result_add = reduce(add_nums, list_1)
print(result_add)

result = 0

for num in list_1:
    result+= num
print(result)

30
30


#### Using Lambda's with Reduce()

In [ ]:
reduce_lamb = reduce

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [85]:
from functools import reduce
list1 = [1, 2, 3]

result1 = reduce((lambda x, y: x * y), list1)
print(result1)




6


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [89]:
def add_nums(num):
    #first we want to set our base case 
    if num <= 1:
        print("add_nums(1) = 1")
        return num
    else:
        print(f"add_nums{num} = {num} + add_nums{num - 1}")
        return num + add_nums(num - 1)
    
add_nums(5)

add_nums5 = 5 + add_nums4
add_nums4 = 4 + add_nums3
add_nums3 = 3 + add_nums2
add_nums2 = 2 + add_nums1
add_nums(1) = 1


15

#### Writing a Factorial Function

In [90]:
def factorial(num):
    #set or base case
    if num <= 1:
        return 1
    
    else:
        return num * factorial(num - 1)
    
print(factorial(5))

120


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [93]:
#return num - sub_nums(num - 1)

def sub_nums(num):
    #set our base case
    if num<=1:
        print("sub_nums(1) = 1")
        return num
    
    else:
        print(f"sub_nums({num}) = {num} - sub_nums({num -1})") 
        return num - sub_nums(num -1)
print(sub_nums(5))

sub_nums(5) = 5 - sub_nums(4)
sub_nums(4) = 4 - sub_nums(3)
sub_nums(3) = 3 - sub_nums(2)
sub_nums(2) = 2 - sub_nums(1)
sub_nums(1) = 1
3


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [95]:
def my_range(stop, start, step = 2):
    while start < stop: 
        yield start #YIELD KEYWORD IS WHAT MAKES IT GENERATOR
        start += step
        
for i in my_range(20,2):
    my_generator_value = i
    print(my_generator_value)

2
4
6
8
10
12
14
16
18


#### Infinite Generator

In [99]:
# bad, never create infinite loops

def bad_range(stop, start, step=1):
    while start < stop:
        yield start 
        #make sure you always increment your generators!!!!!! So it doesn't go into an infinite while loop
for i in bad_range(10,1):
    print(i)
    break

1


#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [110]:
def my_range(stop, start, step = 1 ):
    while start < stop: 
        yield start 
        start += step
        
for i in my_range(25,0):
    my_generator_value = i**2
    print(my_generator_value)

0
1
4
9
16
25
36
49
64
81
100
121
144
169
196
225
256
289
324
361
400
441
484
529
576


In [111]:
def squared_gen(num):
    while num >= 0:
        yield num ** 2
        num -= 1
        
print([i for i in squared_gen(10)])

[100, 81, 64, 49, 36, 25, 16, 9, 4, 1, 0]


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [144]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

[x for x in places if x.strip()]

['Argentina', 'San Diego', 'Boston', 'New York']

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [150]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key=lambda name: name.lower().split(" ")[-1])
print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [172]:
# F = (9/5)*C + 32
#with a map you can have an x and y variable and do two different things with them
#map(lambda x, y: x(something to x), y (something to y)))
#return as a tuple
#list(filter(function, iterable(list, dict, tuples, set, etc)))



#my_list = [2, 4, 6, 8, 10, 12, 14, 16]
places = [('Nashua', 32),("Boston", 12),("Los Angelos", 44),("Miami", 29)]

#new_list = list(map(lambda x: x * 2 - 1 , my_list))
newplaces = list(map(lambda x: (x[0], x[1] * (float(9/5) + 32), places)))

#print(new_list)
print(newplaces)




TypeError: map() must have at least two arguments.

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [173]:
#videos on recursion/fib seq

def fib(i):
    if i > 1:
        return fib(i-1) + fib(i-2)
    else:
        return i
for num in range(6):
    print(num, ":", fib(num))

0 : 0
1 : 1
2 : 1
3 : 2
4 : 3
5 : 5
